File: test-persistent/module.go
package main

import (
	"bytes"
	"strconv"

	"github.com/pantopic/wazero-state-machine/sdk-go"
)

var (
	idx   uint64
	items uint64
	sets  uint64
)

func main() {
	statemachine.RegisterPersistent(open, update, finish, read)
	statemachine.Streamable(streamOpen, streamRecv, streamClosed)
	statemachine.Watchable(watchOpen, watchClosed)
}

func open() uint64 {
	return 0
}

func update(index uint64, cmd []byte) (value uint64, data []byte) {
	items++
	value = idx
	idx = index
	data = bytes.ReplaceAll(cmd, []byte(`test`), []byte(`best`))
	return
}

func finish() {
	sets++
}

var (
	queryIndex = []byte(`index`)
	queryItems = []byte(`items`)
	querySets  = []byte(`sets`)
)

func read(query []byte) (value uint64, data []byte) {
	if bytes.Equal(query, queryIndex) {
		value = idx
	} else if bytes.Equal(query, queryItems) {
		value = items
	} else if bytes.Equal(query, querySets) {
		value = sets
	} else {
		panic(`Unrecognized query: "` + string(query) + `"`)
	}
	return
}

func streamOpen() {
	println(`wasm stream open`)
}

func streamRecv(data []byte) {
	println(`wasm stream recv ` + string(data))
	if bytes.Equal(data, []byte(`close`)) {
		println(`wasm stream close start`)
		statemachine.StreamClose()
		println(`wasm stream close complete`)
	} else {
		println(`wasm stream send start`)
		statemachine.StreamSend(1, data)
		println(`wasm stream send complete`)
	}
}

func streamClosed() {
	println(`wasm stream closed`)
}

func watchOpen(data []byte) {
	println(`wasm watch open`)
	n, err := strconv.Atoi(string(data))
	if err != nil {
		panic(err)
	}
	for i := range n {
		println(`wasm watch send ` + strconv.Itoa(i+1))
		statemachine.WatchSend(1, []byte(strconv.Itoa(i+1)))
	}
	statemachine.WatchClose()
}

func watchClosed() {
	println(`wasm watch closed`)
}

File: test-persistent-zig/module.zig
const std = @import("std");
const statemachine = @import("statemachine");

comptime {
    _ = statemachine;
}

var idx: u64 = 0;
var items: u64 = 0;
var sets: u64 = 0;

export fn _start() void {
    statemachine.persistent(open, update, finish, read);
    statemachine.streaming(streamOpen, streamRecv, streamClosed);
    statemachine.watchable(watchOpen, watchClosed);
}

fn open() u64 {
    return 0;
}

fn update(index: u64, cmd: []u8) statemachine.Result {
    items += 1;
    const value = idx;
    idx = index;
    var i: usize = 0;
    while (std.mem.indexOfPos(u8, cmd, i, "test")) |pos| {
        @memcpy(cmd[pos .. pos + 4], "best");
        i = pos + 4;
    }
    return .{ .value = value, .data = cmd };
}

fn finish() void {
    sets += 1;
}

fn read(query: []u8) statemachine.Result {
    if (std.mem.eql(u8, query, "index")) {
        return .{ .value = idx };
    } else if (std.mem.eql(u8, query, "items")) {
        return .{ .value = items };
    } else if (std.mem.eql(u8, query, "sets")) {
        return .{ .value = sets };
    }
    @panic("Unrecognized query");
}

fn streamOpen() void {
    std.debug.print("wasm stream open\n", .{});
}

fn streamRecv(data: []u8) void {
    std.debug.print("wasm stream recv {s}\n", .{data});
    if (std.mem.eql(u8, data, "close")) {
        std.debug.print("wasm stream close start\n", .{});
        statemachine.streamClose();
        std.debug.print("wasm stream close complete\n", .{});
    } else {
        std.debug.print("wasm stream send start\n", .{});
        statemachine.streamSend(1, data);
        std.debug.print("wasm stream send complete\n", .{});
    }
}

fn streamClosed() void {
    std.debug.print("wasm stream closed\n", .{});
}

fn watchOpen(data: []u8) void {
    std.debug.print("wasm watch open\n", .{});
    const n = std.fmt.parseInt(u64, data, 10) catch @panic("Invalid watch count");
    var i: u64 = 1;
    var tmp: [20]u8 = undefined;
    while (i <= n) : (i += 1) {
        const s = std.fmt.bufPrint(&tmp, "{d}", .{i}) catch unreachable;
        std.debug.print("wasm watch send {s}\n", .{s});
        statemachine.watchSend(1, s);
    }
    statemachine.watchClose();
}

fn watchClosed() void {
    std.debug.print("wasm watch closed\n", .{});
}
